### PR TITLE
[FLINK-12996][table-common] Add type validators and strategies for FLIP-65

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This adds a set of pre-defined type validators and strategies which will be extracted automatically as defined in FLIP-65.

## Brief change log

See commit messages.

## Verifying this change

This change added tests and can be verified as follows:

- `TypeStrategiesTest`
- `InputTypeValidatorsTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/AndTypeArgumentValidator.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Validator that checks for a conjunction of multiple {@link ArgumentTypeValidator}s into one like
+ * {@code f(NUMERIC & STRING)}.
+ */
+@Internal
+public class AndTypeArgumentValidator implements ArgumentTypeValidator {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/AndTypeArgumentValidator.java]:
```suggestion
public final class AndTypeArgumentValidator implements ArgumentTypeValidator {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/LiteralTypeValidator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Validator that checks if a single argument is a literal.
+ */
+@Internal
+public class LiteralTypeValidator implements ArgumentTypeValidator, InputTypeValidator {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/LiteralTypeValidator.java]:
```suggestion
public final class LiteralTypeValidator implements ArgumentTypeValidator, InputTypeValidator {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/OrTypeArgumentValidator.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Validator that checks for a disjunction of multiple {@link ArgumentTypeValidator}s into one like
+ * {@code f(NUMERIC | STRING)}.
+ */
+@Internal
+public class OrTypeArgumentValidator implements ArgumentTypeValidator {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/OrTypeArgumentValidator.java]:
```suggestion
public final class OrTypeArgumentValidator implements ArgumentTypeValidator {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/OrTypeInputValidator.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Validator that checks for a disjunction of multiple {@link InputTypeValidator}s into one like
+ * {@code f(NUMERIC) || f(STRING)}.
+ */
+@Internal
+public class OrTypeInputValidator implements InputTypeValidator {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/OrTypeInputValidator.java]:
```suggestion
public final class OrTypeInputValidator implements InputTypeValidator {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/SequenceInputValidator.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Validator that checks for a sequence of {@link ArgumentTypeValidator}s for validating an entire
+ * function signature like {@code f(STRING, NUMERIC)}.
+ */
+@Internal
+public class SequenceInputValidator implements InputTypeValidator {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/SequenceInputValidator.java]:
```suggestion
public final class SequenceInputValidator implements InputTypeValidator {
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/SequenceInputValidator.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			}
+		}
+		return true;
+	}
+
+	@Override
+	public List<Signature> getExpectedSignatures(FunctionDefinition definition) {
+		final List<Signature.Argument> arguments = new ArrayList<>();
+		// according to precondition we can assume single type validators
+		// thus we can pass constant 0s
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/validators/SequenceInputValidator.java]: remove the comment

**Comment by twalthr** [comment]
Thanks for the improvements @dawidwys. Sometimes one has to think out of the box :D I fixed a couple of bugs but the PR should be good to merge now.

**Pull request closed without merging**